Wms support

peppsac opened the pull request:
WMS support based on the work done by @qdnguyen 

Depends on PR #149

qdnguyen committed:
Add example WMS layer

qdnguyen committed:
Add WMS support

Fix #92

qdnguyen commented:
IO_driver selection in providers should be more generic?  #157

peppsac commented:
PR ready for review.

gchoqueux commented:
OK i review

gchoqueux commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -30,39 +32,235 @@ define('Core/Commander/Providers/WMS_Provider', [
-            this.srs = options.srs || "";
-            this.width = defaultValue(options.width, 256);
-            this.height = defaultValue(options.height, 256);
+            this._ready       = false;
```
`this._ready` is unused

qdnguyen replied:
this variable was deleted in commit 0c23027

gchoqueux commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -30,39 +32,235 @@ define('Core/Commander/Providers/WMS_Provider', [
+            //var bbox = (coord.minCarto.longitude - Math.PI)* 180 / Math.PI + "," + coord.minCarto.latitude* 180 / Math.PI + "," +
+            //           (coord.maxCarto.longitude - Math.PI )*180/ Math.PI + "," + coord.maxCarto.latitude* 180 / Math.PI;
+
+            var bbox = coord.minCarto.latitude * 180.0 / Math.PI +
```
The coord to BBox must be put into the Core/Geographic
or
Understand, why coord are incorrect.

qdnguyen replied:
I dont understand what you mean but this coord is given by tileProvider and i transform it to map center which is origin of wms layers.

gchoqueux replied:
``` js
var bbox = coord.minCarto.latitude * 180.0 / Math.PI +
"," +
(coord.minCarto.longitude - Math.PI)* 180.0 / Math.PI +
","+
coord.maxCarto.latitude* 180.0 / Math.PI +
"," +
(coord.maxCarto.longitude - Math.PI )*180.0 / Math.PI;
```

This part must to move in Core and the question is 'Why we need to convert longitude `(coord.maxCarto.longitude - Math.PI )`'. May be we should correct longitude in tile

qdnguyen replied:
As i said, origin of wms layers is at map center, because in WGS84 the prime meridian pass through GreenWich, which is not the case in our itowns by now. I have open an issue some months ago but it's not resolved yet. This problem touch to every module, so for now we can not move it to core, but one time core is modified, we can delete this code. #19

peppsac replied:
If issue #19 is the sole reason why you transform `longitude` please add a comment in the code:

``` js
// shifting longitude because of issue #19
```

And please remove dead code.

qdnguyen replied:
done!

peppsac replied:
did you push your change? (I don't see it)

gchoqueux commented:
Good Feature
- `WMS_Provider` and `WMTS_Provider` will need to factor. there is a lot of redundancy between the two objects. We can imagine RasterDataTileProvider.

gchoqueux commented on src/Core/Math/Rectangle.js:
```diff
@@ -3,14 +3,61 @@
  * Class: Rectangle
  */
 
-function Rectangle() {
```
`Scene/BoundingBox` is very similar to `Core/Math/Rectangle`
It needs to merge in one object

qdnguyen replied:
rectangle is used in region check with only 4 coords: west, east, south, north. It's faster than bbox check. 
From rectangle we can have the bbox in 2D but in theory rectangle is not bbox. I refer to keep them separately to make code more clear.

gchoqueux commented:
Why there are `urlGlobalIR` and `urlClouds` in WMS
must be removed and replace by custom `WMS_Provider` like WMTS_Provider's example with openStreetMap

qdnguyen commented:
ok, i have cleaned urlGlobalIR and urlClouds, we can request image from these servers with our new wms_provider.

qdnguyen commented:
> WMS_Provider and WMTS_Provider will need to factor. there is a lot of redundancy between the two objects. We can imagine RasterDataTileProvider.

Instead of RasterDataTileProvider, we can also select io_driver according to input info when layer is added? so that providers contain only an instance of io_driver? regardless of data format? But this work should be done after merged because it touch to many providers, not only wms_provider.

gchoqueux commented:
there is a display artifact  (because of the lens texture)
Can you remove it?

![capture du 2016-07-19 11-43-54](https://cloud.githubusercontent.com/assets/11291849/16946460/f1d03b82-4da9-11e6-882a-945fb922d258.png)

qdnguyen commented:
i dont know where it comes from, but it's not from wms module!

gchoqueux commented:
> i dont know where it comes from, but it's not from wms module!

->

> because of the lens texture

There isn't this bug in master

gchoqueux commented:
There's white line on equator
![capture du 2016-07-21 13-51-44](https://cloud.githubusercontent.com/assets/11291849/17021806/a52f1828-4f4a-11e6-83e0-5c3a8bf6aa64.png)

peppsac committed:
Fix rendering artifacts

* make sure that gl_FragColor.a always = 1.0, because we never
want to see through

* only increment valid texture count if layerColor.a is > 0. Fix
white line at the edge of WMS bounding-box

peppsac commented:
I pushed a fix for the lensflare issue.

peppsac commented:
I fixed the white line bug and  have rebased the branch on master.
@qdnguyen do you mind if I force push this branch?

qdnguyen commented:
yes, do it, thanks!

peppsac commented:
Done.
@gchoqueux is there something left that prevents merging this branch?

Jeremy-Gaillard commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -30,71 +32,221 @@ define('Core/Commander/Providers/WMS_Provider', [
-            this.height = defaultValue(options.height, 256);
+            this.getTextureFloat;
+
+            if(this.support)
```
this.support is always undefined.

gchoqueux commented:
There's a problem with wms textures?
![capture du 2016-07-26 12-08-06](https://cloud.githubusercontent.com/assets/11291849/17134254/cfc5c40a-5329-11e6-9a38-ef48b865e397.png)

peppsac commented:
It's an error handling issue, and it's not specific to WMS.
The example WMS server returns an error sometimes, and itowns doesn't handle this very well atm.

peppsac commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -30,71 +32,221 @@ define('Core/Commander/Providers/WMS_Provider', [
+            return bboxRegion.intersect(bbox);
+        };
+
+        WMS_Provider.prototype.getColorTextures = function(tile, layer) {
```
This function is not needed, please get rid of it.

qdnguyen replied:
it is needed in preprocessLayer function of ApiGlobe module

peppsac replied:
I meant the `getColorTextures` function

qdnguyen replied:
for data selection in providers, i have created an issue #157 but we must do it right now in this PR? because we must do for all providers, not only wms_provider.

peppsac commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -30,71 +32,221 @@ define('Core/Commander/Providers/WMS_Provider', [
+            //     return Promise.resolve(texture);
+            // }
+            // -> bug #74
+
```
Remove dead code please.

peppsac commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -30,71 +32,221 @@ define('Core/Commander/Providers/WMS_Provider', [
+            var destination = command.paramsFunction.destination;
+            var tile = command.requester;
+
+            if(destination === 1) {
```
IMHO the test should depend solely on `layer.mimetype`:
- if it's a color image format (eg: png, jpg) use `getColorTexture`
- if it's an `xbil` format then use `getElevationTexture` (which should probably have a different name).

qdnguyen replied:
Issue #157

peppsac commented:
Rebased on master

peppsac committed:
Ignore error from IoDriver_Image

This is obviously not the correct way to handle errors,
but now that PR #169 has landed, /this/ is undefined so
this.src causes an error.

peppsac commented:
And added a workaround for an error in IoDriver_Image.
@gchoqueux : r?

qdnguyen commented:
@peppsac  did you rebase this PR on master? i did clean up some codes but i have not commited yet.

peppsac commented:
Yes it's rebased.

qdnguyen commented:
ok, it was not a big thing, please push it :)

peppsac committed:
Remove 'destination' knowledge from providers

Instead use layer's mimetype to choose the correct driver and
computation to perform.

See issue #157

tbroyer commented on src/Core/Commander/Providers/IoDriver_Image.js:
```diff
@@ -35,11 +35,9 @@ IoDriver_Image.prototype.read = function(url) {
         image.addEventListener('error', function( /*event*/ ) {
-
-            this.src = '';
+            // TODO: real error managament
```
See #170

tbroyer commented on index.html:
```diff
@@ -384,6 +384,27 @@
         // });
 
+        itowns.viewer.addImageryLayer({
+                                        url       : "https://wxs.ign.fr/gratuit/geoportail/v/wms?",
```
`WMS_Provider` uses `layer.url + "?SERVICE=WMS…"` which results in a doubled `??` in the URL. That doesn't seem to cause problems to the server but is probably not what was intended.

I don't know WMS enough but either the `?` here should be removed, or `WMS_Provider` should look at whether `layer.url` contains a `?` already and conditionally append with a `?` or a `&` (or possibly no separator if `layer.url` ends with one of those)

gchoqueux commented:
All levels of WMS are loaded. WMTS uses ancestors, it isn't the same things for WMS?
It's better to load only the necessary levels.

gchoqueux commented:
What is the policy to display WMS in function of level's tile?
It's possible to set this rule?

gchoqueux commented:
If you add only wms in index.html, the navigator don't stop to request WMS images.

qdnguyen commented:
the notion of level is not exist in wms, the service generates the level according to bbox, i think we can limit LODs by checking bbox size but i'm not sur about this, may be @gcebelieu or @elias75015 can reply? Anyway, without getcapabilities, we have less infos!

p/s: the downsload is constrained by checking west, east, south, north but i dont think it's enough to limit LODs

gchoqueux commented:
in 'Globe Loaded' Event, some requests WMS are launched and nothing is displayed. why?
All queries fail, it's possible to test before launch a request?

gchoqueux commented:
> the notion of level is not exist in wms

Yes, it's true
But the WMS images are linked with tiles and these tiles have a level.
WMS_provider cuts in tiles and it's the same WMTS's mechanism.
It must be possible to use the textures's ancestors

qdnguyen commented:
> But the WMS images are linked with tiles and these tiles have a level.
> WMS_provider cuts in tiles and it's the same WMTS's mechanism.

For now, wms depends on wmts tile bbox, but in general, you can request it with any bbox which will be taken into account by wms to generate LODs. 

> It must be possible to use the textures's ancestors

what do you mean 'texture's ancestors'? is it a quadtree? yes, we can create a quadtree for wms which cut tile with a specifique tilingScheme. And maxLevel, minLevel of quadtree can be defined by getcap?

qdnguyen commented:
> in 'Globe Loaded' some requests WMS are launched and nothing is displayed. why?

I dont understand your question, wms layers are added through  itowns.viewer.addImageryLayer, not addEventListener 'Globe Loaded'

peppsac commented:
Regarding,

> If you add only wms in index.html, the navigator don't stop to request WMS images.

and 

> What is the policy to display WMS in function of level's tile?

Actually providers shouldn't be aware of `level` at all - `NodeProcess` should probably take care of this.

And the issue you mention is probably caused by this very issue, as this code is used to determine if we should download another texture:

``` js
var level = this.Textures[id][0].level;
```

I'll work on another PR to land before this one, to move all this `level` management to `NodeProcess`.
(it's also related to error management, because as you can see only the first texture is tested to determine if we need to refine the color layer)

peppsac committed:
Remove unused 'useParent' method

peppsac committed:
Add WMS/WGS84 pitch transformation function

Compute the pitch to use when assigning a parent's
WMS texture to a node.

 # If applied, this commit will...

peppsac committed:
Remove redundant info in TileMesh/LayeredMaterial

peppsac committed:
Workaround 'getDelta' bug

peppsac committed:
Move textures' level management to NodeProcess

Providers are data structures agnostic so move the level handling to NodeProcess.

This also allows to move the 'use ancestor texture' mechanism in NodeProcess so
it's now used for all providers.

gchoqueux commented:
WMS_SUPPORT2 Warning console
- /src/Scene/NodeProcess.js  307:33  error  'err' is defined but never used  no-unused-vars
- /src/Renderer/LayeredMaterial.js
  - 294:5  error  Unreachable code  no-unreachable
  - 295:9  error  Unreachable code  no-unreachable
  - 297:9  error  Unreachable code  no-unreachable

gchoqueux commented:
> in 'Globe Loaded' some requests WMS are launched and nothing is displayed. why?

In start loading, There's WMS requests. They return transparent texture. Perhaps, it needs to crop bbox's request with global bbox. I will investigate on the subject

peppsac commented:
> In start loading, There's WMS requests. They return transparent texture. Perhaps, it needs to crop bbox's request with global bbox.

The bbox is big because it covers France + overseas territory.
But the server returns non-transparent texture only for smaller bbox (level 4 or 5)

gchoqueux commented:
> But the server returns non-transparent texture only for smaller bbox (level 4 or 5)

The `MinScaleDenominator` and `MaxScaleDenominator` elements define the range of scales for which it is appropriate to generate a map of a Layer. 

But there's no info to scale denominator in GetCapabilities.

gchoqueux commented:
Nice : 
- WMS don't load all level texture.
- On the Start, there's no loop requesting

gchoqueux commented:
`TileMesh.prototype.changeSequenceLayers` doesn't work anymore. `sequence` is undefined.

peppsac commented:
Branch updated

gchoqueux commented:
r+
I create a new issue for 'scale denominator'
